### PR TITLE
fix(blood pacts): Correct physical blood pact damage calculation

### PR DESCRIPTION
**Pull request by davismj**
A lot of the info out there quotes a forum post from allakazham that has since been deleted. The author was presumably kegsay and he wrote a lot of the same findings in his livejournal here: https://kegsay.livejournal.com/tag/smn!. This pr implements those findings more faithfully.

Specifically:

- removed lv diff calculation, does not apply https://kegsay.livejournal.com/5510.html
- removed fstr cap https://kegsay.livejournal.com/5510.html
- added fstr_min and fstr_max calculations https://kegsay.livejournal.com/5510.html
- allow crits on all hits of multi-hit blood pact
- updated M/S values on chaotic strike, crescent fang, eclipse bite https://kegsay.livejournal.com/4169.html
- updated accuracy bonus on evokers spats, summoners bracers to +14 https://kegsay.livejournal.com/4471.html

All of the above are found in Kegsay's posts except the crit change. The current code prevented crits on subsequent hits completely, and I couldn't find any justification for this, so I removed it.

Please note that wiki has information from Kegsay's research copy pasted out of context. Take for example this, found on this page: https://ffxiclopedia.fandom.com/wiki/Calculating_Blood_Pact_Damage

```
Celestial Avatars + Fenrir
BASE DAMAGE = (AVATARS LEVEL * 0.74) - MONSTER VIT/4

Carbuncle
BASE DAMAGE = (AVATARS LEVEL * 2/3) - MONSTER VIT/4
```

The original text reads:

```
For the sake of purpose, a few generalisations are going to be made in order to provide a rough estimation of how much base damage your avatar has:

Avatar base damage = (AVATAR LEVEL x 0.74) - (Monster VIT/4)
```

In other words, this wasn't a formula, but an estimate for testing purposes.

### CODE REVIEW

File: scripts/globals/summon.lua [hunk 1]
```diff
@@ -1,241 +1,187 @@
...
-require("scripts/globals/common");
-require("scripts/globals/status");
-require("scripts/globals/msg");
+require("scripts/globals/common")
+require("scripts/globals/status")
+require("scripts/globals/msg")
 
-function AvatarPhysicalMove(avatar,target,skill,numberofhits,accmod,dmgmod1,dmgmodsubsequent,tpeffect,mtp100,mtp200,mtp300)
-    returninfo = {};
+function getSummoningSkillOverCap(summoner)
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Does SMN as a subjob get this benefit or is it always 0? If it's always 0, this code is fine, otherwise subjob needs to be handled.

File: scripts/globals/summon.lua [hunk 2]
```diff
@@ -1,241 +1,187 @@
...
+        minRatio = 1/3
+    elseif (ratio < 1.6) then
+        maxRatio = ((2/3) * ratio) + (1/3)
+        minRatio = ((7/9) * ratio) - (4/9)
+    elseif (ratio <= 1.8) then
+        maxRatio = 1.8
+        minRatio = 1
+    elseif (ratio < 3.6) then
+        maxRatio = (2.4 * ratio) - 2.52
+        minRatio = ((5/3) * ratio) - 2
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Remove all the parentheses in this block. None of them are necessary. Order of operations handle it correctly. Outer parentheses in conditionals are also against current style-guide.

Also space out numbers and operators.

File: scripts/globals/summon.lua [hunk 3]
```diff
@@ -1,241 +1,187 @@
...
-    if (nativecrit > 0.2) then --caps!
-        nativecrit = 0.2;
-    elseif (nativecrit < 0.05) then
-        nativecrit = 0.05;
+    local hitsdone = 1
+    local hitslanded = 0
+    local hitdmg = 0
+    local finaldmg = 0
+
+    if math.random() <= hitrate then
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: math.random() can return 0 so this should be a < instead of <=

File: scripts/globals/summon.lua [hunk 4]
```diff
@@ -1,241 +1,187 @@
...
-            pdif = pdif/1000;
-            finaldmg = finaldmg + subsequenthitdamage * pdif;
-            hitslanded = hitslanded + 1;
-        end
-        hitsdone = hitsdone + 1;
+        hitsdone = hitsdone + 1
     end
 
     -- all hits missed
     if (hitslanded == 0 or finaldmg == 0) then
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Rip out these parentheses.

File: scripts/globals/summon.lua [hunk 5]
```diff
@@ -1,241 +1,187 @@
...
-    elseif (dSTR >= -7) then
-        fSTR2 = ((dSTR+9)/2);
-    elseif (dSTR >= -15) then
-        fSTR2 = ((dSTR+10)/2);
-    elseif (dSTR >= -21) then
-        fSTR2 = ((dSTR+12)/2);
-    else
-        fSTR2 = ((dSTR+13)/2);
+    -- apply ftp bonus
+    if (tpeffect == TP_DMG_BONUS) then
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: No parentheses

File: scripts/globals/summon.lua [hunk 6]
```diff
@@ -1,241 +1,187 @@
...
+    returninfo.hitslanded = hitslanded
+
+    return returninfo
+end
+
+-- minFstr = dSTR/4 + 0.5
+-- maxFstr = dSTR/4 + 0.25
+function avatarFSTR(str, def_vit)
+    local dSTR = atk_str - def_vit
+    return math.floor(dSTR/4 + 0.5), math.floor(dSTR/4 + 0.25)
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Space out operators

File: scripts/globals/summon.lua [hunk 7]
```diff
@@ -1,241 +1,187 @@
...
+end
+
+function avatarHitDmg(dmg, pdifMin, pdifMax, fstrMin, fstrMax, critrate)
+    local fstr = math.random(fstrMin, fstrMax)
+    local pdif = math.random(pdifMin * 1000, pdifMax * 1000) / 1000
+    if math.random() <= critrate then
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: < instead of <=

File: scripts/globals/summon.lua [hunk 8]
```diff
@@ -1,241 +1,187 @@
...
     end
-    return fSTR2;
-end;
+    return (dmg + fstr) * pdif
+end
 
 function AvatarFinalAdjustments(dmg,mob,skill,target,skilltype,skillparam,shadowbehav)
 
     -- physical attack missed, skip rest
     if (skilltype == dsp.attackType.PHYSICAL and dmg == 0) then
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Gut parentheses

File: scripts/globals/summon.lua [hunk 9]
```diff
@@ -1,241 +1,187 @@
...
+        return 0
     end
 
     -- set message to damage
     -- this is for AoE because its only set once
-    skill:setMsg(dsp.msg.basic.DAMAGE);
+    skill:setMsg(dsp.msg.basic.DAMAGE)
 
     --Handle shadows depending on shadow behaviour / skilltype
     if (shadowbehav < 5 and shadowbehav ~= MOBPARAM_IGNORE_SHADOWS) then --remove 'shadowbehav' shadows.
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Boo Parentheses. Just gonna have you style the whole file since you're here. :)

File: scripts/globals/summon.lua [hunk 10]
```diff
@@ -1,241 +1,187 @@
...
-        targShadows = target:getMod(dsp.mod.UTSUSEMI);
-        shadowType = dsp.mod.UTSUSEMI;
+        targShadows = target:getMod(dsp.mod.UTSUSEMI)
+        shadowType = dsp.mod.UTSUSEMI
         if (targShadows==0) then --try blink, as utsusemi always overwrites blink this is okay
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Parenth and space out operator

File: scripts/globals/summon.lua [hunk 11]
```diff
@@ -1,241 +1,187 @@
...
-            targShadows = target:getMod(dsp.mod.BLINK);
-            shadowType = dsp.mod.BLINK;
+            targShadows = target:getMod(dsp.mod.BLINK)
+            shadowType = dsp.mod.BLINK
         end
 
         if (targShadows>0) then
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: same as above

File: scripts/globals/summon.lua [hunk 12]
```diff
@@ -1,241 +1,187 @@
...
         -- Blink has a VERY high chance of blocking tp moves, so im assuming its 100% because its easier!
             if (targShadows >= shadowbehav) then --no damage, just suck the shadows
-                skill:setMsg(dsp.msg.basic.SHADOW_ABSORB);
-                target:setMod(shadowType,(targShadows-shadowbehav));
+                skill:setMsg(dsp.msg.basic.SHADOW_ABSORB)
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Gut parenth in all these conditionals
Don't need to surround the math in parentheses either (I know that was there before and wasn't you)

File: scripts/globals/summon.lua [hunk 13]
```diff
@@ -1,241 +1,187 @@
...
                 end
-                return shadowbehav;
+                return shadowbehav
             else -- less shadows than this move will take, remove all and factor damage down
-                dmg = dmg * ((shadowbehav-targShadows)/shadowbehav);
-                target:setMod(dsp.mod.UTSUSEMI,0);
-                target:setMod(dsp.mod.BLINK,0);
-                target:delStatusEffect(dsp.effect.COPY_IMAGE);
-                target:delStatusEffect(dsp.effect.BLINK);
+                dmg = dmg * ((shadowbehav-targShadows)/shadowbehav)
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Space out the operators and the outside set of parentheses are unneeded. (Keep the ones around the subtraction)

File: scripts/globals/summon.lua [hunk 14]
```diff
@@ -1,241 +1,187 @@
...
+                target:setMod(dsp.mod.UTSUSEMI,0)
+                target:setMod(dsp.mod.BLINK,0)
+                target:delStatusEffect(dsp.effect.COPY_IMAGE)
+                target:delStatusEffect(dsp.effect.BLINK)
             end
         end
     elseif (shadowbehav == MOBPARAM_WIPE_SHADOWS) then --take em all!
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Last time I'll label parentheses for conditionals. Just continue gutting them, pretty please.

File: scripts/globals/summon.lua [hunk 15]
```diff
@@ -1,241 +1,187 @@
...
             if (prevAnt == 0) then
                 --100% proc
-                teye:setPower(1);
-                skill:setMsg(dsp.msg.basic.ANTICIPATE);
-                return 0;
+                teye:setPower(1)
+                skill:setMsg(dsp.msg.basic.ANTICIPATE)
+                return 0
             end
             if ( (math.random()*100) < (80-(prevAnt*10)) ) then
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Space out the operators and ALL of these parentheses can go.

File: scripts/globals/summon.lua [hunk 16]
```diff
@@ -244,85 +190,85 @@ function AvatarFinalAdjustments(dmg,mob,skill,target,skilltype,skillparam,shadow
...
     -- handle elemental resistence
     if (skilltype == dsp.attackType.MAGICAL and target:hasStatusEffect(dsp.effect.MAGIC_SHIELD)) then
-        return 0;
+        return 0
     end
 
     -- handling phalanx
-    dmg = dmg - target:getMod(dsp.mod.PHALANX);
+    dmg = dmg - target:getMod(dsp.mod.PHALANX)
     if (dmg<0) then
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Space this boy out

File: scripts/globals/summon.lua [hunk 17]
```diff
@@ -244,85 +190,85 @@ function AvatarFinalAdjustments(dmg,mob,skill,target,skilltype,skillparam,shadow
...
-        return 0;
+        return 0
     end
 
     --handle invincible
     if (target:hasStatusEffect(dsp.effect.INVINCIBLE) and skilltype==dsp.attackType.PHYSICAL) then
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: More spacing

File: scripts/globals/summon.lua [hunk 18]
```diff
@@ -244,85 +190,85 @@ function AvatarFinalAdjustments(dmg,mob,skill,target,skilltype,skillparam,shadow
...
-        return 0;
+        return 0
     end
     -- handle pd
     if ((target:hasStatusEffect(dsp.effect.PERFECT_DODGE) or target:hasStatusEffect(dsp.effect.ALL_MISS) )
             and skilltype==dsp.attackType.PHYSICAL) then
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: More spaci-- you get the idea

File: scripts/globals/summon.lua [hunk 19]
```diff
@@ -244,85 +190,85 @@ function AvatarFinalAdjustments(dmg,mob,skill,target,skilltype,skillparam,shadow
...
+        return 0
     end
-    
+
     -- Calculate Blood Pact Damage before stoneskin
     dmg = dmg + dmg * mob:getMod(dsp.mod.BP_DAMAGE) / 100
 
     -- handling stoneskin
-    skin = target:getMod(dsp.mod.STONESKIN);
+    skin = target:getMod(dsp.mod.STONESKIN)
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: You can actually use utils.stoneskin here

File: scripts/globals/summon.lua [hunk 20]
```diff
@@ -244,85 +190,85 @@ function AvatarFinalAdjustments(dmg,mob,skill,target,skilltype,skillparam,shadow
...
+end
 
 function avatarFTP(tp,ftp1,ftp2,ftp3)
     if (tp < 1000) then
-        tp = 1000;
+        tp = 1000
     end
     if (tp >= 1000 and tp < 2000) then
-        return ftp1 + ( ((ftp2-ftp1)/100) * (tp-1000));
+        return ftp1 + ( ((ftp2-ftp1)/100) * (tp-1000))
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Holy parentheses, batman. Let order of operations do it's job.

`ftp1 + (ftp2 - ftp1) / 100 * (tp - 1000)`

File: scripts/globals/summon.lua [hunk 21]
```diff
@@ -244,85 +190,85 @@ function AvatarFinalAdjustments(dmg,mob,skill,target,skilltype,skillparam,shadow
...
     elseif (tp >= 2000 and tp <= 3000) then
         -- generate a straight line between ftp2 and ftp3 and find point @ tp
-        return ftp2 + ( ((ftp3-ftp2)/100) * (tp-2000));
+        return ftp2 + ( ((ftp3-ftp2)/100) * (tp-2000))
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: `ftp2 + (ftp3 - ftp2) / 1000 * (tp - 2000)`

File: scripts/globals/summon.lua [hunk 22]
```diff
@@ -244,85 +190,85 @@ function AvatarFinalAdjustments(dmg,mob,skill,target,skilltype,skillparam,shadow
...
+    return 1 -- no ftp mod
+end
 
 -- Checks if the summoner is in a Trial Size Avatar Mini Fight (used to restrict summoning while in bcnm)
 function avatarMiniFightCheck(caster)
-    local result = 0;
-    local bcnmid;
+    local result = 0
+    local bcnmid
     if (caster:hasStatusEffect(dsp.effect.BATTLEFIELD) == true) then
```
Comment (nfaletra) [inline scripts/globals/summon.lua]: Get rid of `== true`

File: sql/item_mods_pet.sql [hunk 23]
```diff
@@ -114,7 +114,7 @@ INSERT INTO `item_mods_pet` VALUES (13975,27,-2,1);     -- Evoker's Bracers Avat
 INSERT INTO `item_mods_pet` VALUES (14103,27,-2,1);     -- Evoker's Pigaches Avatar: Enmity -2
 INSERT INTO `item_mods_pet` VALUES (14103,68,5,1);      -- Avatar: Enhances evasion +5 (?)
 INSERT INTO `item_mods_pet` VALUES (14227,3,10,2);      -- Drachen Brais Wyvern: HP+10%
-INSERT INTO `item_mods_pet` VALUES (14228,25,10,1);     -- Evoker's Spats Avatar: Enhances accuracy +10 (?)
+INSERT INTO `item_mods_pet` VALUES (14228,25,14,1);     -- Evoker's Spats Avatar: Enhances accuracy +14
```
Comment (nfaletra) [inline sql/item_mods_pet.sql]: Where'd you pull this stat from?

afaik it should be +10 and was checking in retail using /checkparam

File: sql/item_mods_pet.sql [hunk 24]
```diff
@@ -135,7 +135,7 @@ INSERT INTO `item_mods_pet` VALUES (14958,63,5,0);      -- Beast Bazubands Pet:
 INSERT INTO `item_mods_pet` VALUES (15031,384,300,3);    -- Pantin Dastanas Automaton: Haste +3%
 INSERT INTO `item_mods_pet` VALUES (15032,384,300,3);    -- Pantin Dastanas +1 Automaton: Haste +3%
 INSERT INTO `item_mods_pet` VALUES (15101,165,3,1);     -- Summoner's Dblt. Avatar: Critical Hit Rate +3%
-INSERT INTO `item_mods_pet` VALUES (15116,25,7,1);      -- Summoner's Brcr. Avatar: Enhances accuracy +7 (?)
+INSERT INTO `item_mods_pet` VALUES (15116,25,14,1);      -- Summoner's Brcr. Avatar: Enhances accuracy +14
```
Comment (nfaletra) [inline sql/item_mods_pet.sql]: Same with this one. Unsure where you pulled this value.

File: sql/item_mods_pet.sql (hunk 23 above)
Comment (davismj) [reply sql/item_mods_pet.sql]: https://kegsay.livejournal.com/4471.html

Specifically lists the item and the value. I can revert it if it is confirmed that this has since changed.

Comment (zynjec) [reply sql/item_mods_pet.sql]: Please be careful with the age of your information.

Newer sources claim +10 accuracy using /checkparams on retail.

Comment (davismj) [reply sql/item_mods_pet.sql]: I believe it but I couldn't find them. Spent quite a bit of time checking for attack, accuracy, and evasion both in en and jp sources.

Comment (zynjec) [reply sql/item_mods_pet.sql]: Both BGWiki and FFXIclopedia state +10 state with /checkparam, which is much more accurate than work that is essentially guessing by comparison.

Comment (davismj) [reply sql/item_mods_pet.sql]: You're right. I was checking for https://ffxiclopedia.fandom.com/wiki/Summoner's_Bracers, which has no info. Went looking for it, found kegsay, and he mentioned spats as well so I changed it as well. Thx.

File: scripts/globals/summon.lua (hunk 1 above)
Comment (davismj) [reply scripts/globals/summon.lua]: nah i introduce a bug, nice catch thanks

File: sql/item_mods_pet.sql (hunk 24 above)
Comment (davismj) [reply sql/item_mods_pet.sql]: https://kegsay.livejournal.com/4471.html

Couldn't find an updated value for this.

Comment (zynjec) [reply sql/item_mods_pet.sql]: https://www.bg-wiki.com/bg/Summoner%27s_Brcr. lists it as +7

Comment (davismj) [reply sql/item_mods_pet.sql]: touche

**Comment by nfaletra** [comment]
Sorry reviewing took so long. I wanted to go through the blog you posted. Just now had some time to really dive into it. Overall, good job.

File: scripts/globals/summon.lua [hunk 25]
```diff
@@ -1,328 +1,258 @@
...
+require("scripts/globals/common")
+require("scripts/globals/status")
+require("scripts/globals/msg")
 
-function AvatarPhysicalMove(avatar,target,skill,numberofhits,accmod,dmgmod1,dmgmodsubsequent,tpeffect,mtp100,mtp200,mtp300)
-    returninfo = {};
+function getSummoningSkillOverCap(avatar)
+    local summoner = avatar:getMaster()
+    local summoningSkill = summoner:getSkillLevel(dsp.skill.SUMMONING_MAGIC)
+    local maxSkill = summoner:getMaxSkillLevel(avatar:getMainLvl(), dsp.job.SMN, dsp.skill.SUMMONING_MAGIC)
```
[the remark this answers is not in the document]
Comment (davismj) [reply scripts/globals/summon.lua]: actually, this seems to work. the self argument isn't used at all.

[the remark this answers is not in the document]
Comment (nfaletra) [reply scripts/globals/summon.lua]: Wow. I just noticed it's using the avatar's level. I'm an idiot. Carry on.

Also wtf with this function on the core side. . .accessing from the top of the stack.

[the remark this answers is not in the document]
Comment (davismj) [reply scripts/globals/summon.lua]: right?
